Improve CSV to json parser for storage

Piranias opened the pull request:
Fix #75, Fix #238 

**Changes proposed in this pull request**:
- in A1 each storage column is parsed separately and only specific column parameters are added to the json
- in C0/D1/E0/E1 the new storage column names are adapted
- change storage_01.csv and json

The following steps were realized, as well (if applies):
- [x] Use in-line comments to explain your code
- [x] Write docstrings to your code
- [x] For new functionalities: Explain in readthedocs
- [ ] Write test(s) for your new patch of code
- [x] Update the CHANGELOG.md
- [x] Apply black (`black . --exclude docs/`)
- [x] Check if benchmark tests pass locally (`pytest tests/test_benchmark.py`)


*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

:x: Check not applicable to this PR

<sub>*For more information on how to contribute check the [CONTRIBUTING.md](https://github.com/rl-institut/mvs_eland/blob/dev/CONTRIBUTING.md).*<sub>

Piranias commented:
the benchmark test dont run through locally. I get the error:
`FAILED tests/test_benchmark.py::test_run_smoothly_csv - FileExistsError: Output folder .../mvs_eland/tests/MVS_outputs_simulation already exists. If you want to overwrite the folder, please chosse overwrite ...`

smartie2076 commented:
> 
> 
> the benchmark test dont run through locally. I get the error:
> `FAILED tests/test_benchmark.py::test_run_smoothly_csv - FileExistsError: Output folder .../mvs_eland/tests/MVS_outputs_simulation already exists. If you want to overwrite the folder, please chosse overwrite ...`
> so this has nothing to do with this PR

Please merge dev branch into this branch - this should be fixed now.

Piranias commented:
> 
> Please merge dev branch into this branch - this should be fixed now.

ok, i think now it works

Bachibouzouk commented on src/D1_model_components.py:
```diff
@@ -315,54 +315,48 @@ def storage_fix(model, dict_asset, **kwargs):
     storage = solph.components.GenericStorage(
         label=dict_asset["label"],
-        existing=dict_asset["capacity"]["installedCap"]["value"],
+        existing=dict_asset["storage capacity"]["installedCap"]["value"],
```
As you use dict_asset["storage capacity"] 3 times, you could create a variable `storage_capacity = dict_asset["storage capacity"]`, it makes the code easier to read :)

Piranias replied:
i didn't write this piece of code, i just renamed the column value. It's the same issue for dict_value['input_power'] and dict_value['output_power']. If you want me to fix the code style in this PR for all of the assets i could do so. But I'm not sure if that makes it more readable because it's always the same structure (dict[asset][subasset] etc.) that makes it easy to read. and this way you see that each time the dict_valus is called.

Bachibouzouk replied:
For me the point was to have shorter lines, less character to read is more readable for me. But I understand it is a personal stand. Don't change it before we have more opinion on it :)

@SabineHaas, @smartie2076 - what do you think?

smartie2076 replied:
I would leave it for now.

Bachibouzouk commented on src/A1_csv_to_json.py:
```diff
@@ -339,10 +340,112 @@ def create_json_from_csv(input_directory, filename, parameters, storage=False):
+                # check if all three columns are available
+                if len(df_copy.columns) < 4 or len(df_copy.columns) > 4:
+                    logging.error(
+                        "The file " + filename + ".csv requires "
```
```suggestion
                        f"The file {filename}.csv requires "
```

Bachibouzouk replied:
In general, could you avoid using "%s" syntax as will be deprecated sooner than the new syntax with f"{filename}" or "{}".format(filename)?

Piranias replied:
ok

Bachibouzouk replied:
That being said, do not make more changes now to the `A1_csv_to_json.py` file, to avoid conflicts with the PR #141

Piranias replied:
ok, good that you say that ;)

Bachibouzouk commented:
@Piranias - I would suggest you postpone the test writing for this feature to after #141 is merged (to avoid conflicts). Are there still blockers for the merge/things to do on this PR?

Piranias commented:
> @Piranias - I would suggest you postpone the test writing for this feature to after #141 is merged (to avoid conflicts). Are there still blockers for the merge/things to do on this PR?

ok, sure. From my point there is no more blockers, only the review is missing.

Bachibouzouk commented on docs/MVS_parameters.rst:
```diff
@@ -127,12 +128,11 @@ energyStorage.csv
 storage_xx.csv
 --------------
-
 **crate**: C-rate is the rate at which the storage can charge or discharge relative to the nominal capacity of the storage.
```
```suggestion
**c_rate**: C-rate is the rate at which the storage can charge or discharge relative to the nominal capacity of the storage.
```

Bachibouzouk commented on src/A1_csv_to_json.py:
```diff
@@ -359,10 +360,112 @@ def create_json_from_csv(input_directory, filename, parameters, storage=False):
         )
-
-    for column in df:
+    df_copy = df
```
```suggestion
    df_copy = df.copy()
```
Otherwise it does not make an actual copy

Bachibouzouk commented on src/A1_csv_to_json.py:
```diff
@@ -359,10 +360,112 @@ def create_json_from_csv(input_directory, filename, parameters, storage=False):
+                if column == "storage capacity":
+                    extra = ["soc_initial", "soc_max", "soc_min"]
+                elif column == "input power" or column == "output power":
+                    extra = ["crate", "opex_var"]
```
```suggestion
                    extra = ["c_rate", "opex_var"]
```

Bachibouzouk commented on src/A1_csv_to_json.py:
```diff
@@ -359,10 +360,112 @@ def create_json_from_csv(input_directory, filename, parameters, storage=False):
+                        # check if not required parameters are set to Nan and
+                        # if not, set them to Nan
+                        if i not in [
+                            "crate",
```
```suggestion
                            "c_rate",
```

Bachibouzouk commented on src/A1_csv_to_json.py:
```diff
@@ -359,10 +360,112 @@ def create_json_from_csv(input_directory, filename, parameters, storage=False):
+    for column in df_copy:
         if column != "unit":
             column_dict = {}
+            # the storage columns are checked for the right parameters,
```
Could you format the following lines with f"{}" instead of "%s" (this part is new so it should not create conflicts :))

Bachibouzouk commented on CHANGELOG.md:
```diff
@@ -53,9 +54,11 @@ tipps for module building, and hint that units in the MVS are not checked (PR #2
 - Fix plotting error in F1, plot only if Data frame is not empty (#230, #234)
 - Fix naming error for storages (#166)
 - Benchmark test that the simulation is running with default settings (#254)
+- Fix specific parameters for each storage column (#75)
```
```suggestion
- Fix specific parameters for each storage column (#259)
```
The convention is to always refer to the PR number, not the issue number

smartie2076 commented on src/A1_csv_to_json.py:
```diff
@@ -328,27 +328,28 @@ def create_json_from_csv(input_directory, filename, parameters, storage=False):
-                    "check %s",
-                    input_directory + " for correct parameter names.",
-                )
+    if storage is False:
```
> if storage is False:

Can you change  `storage` to a variable name that is directly graspable, eg. `asset_is_a_storage` ?

Piranias commented:
> I would describe the change done in the changelog (one entry is for the bugfix, another could be added for the description of the new structure you entered)
> 
> I suggested that you format the strings with f"{}" in the lines that are new anyway (green in the "File changed tab" on github)
> 
> The test for this feature will be implemented after #141 is merged :)

uups sorry, did you mean i should wait for #141 to merge and add the tests before merging this PR? 

I added the line to the changelog and fixed the logging messages as well as the c_rate parameter name.

smartie2076 commented on src/A1_csv_to_json.py:
```diff
@@ -502,23 +576,19 @@ def add_storage_components(storage_filename, input_directory):
+    if not os.path.exists(os.path.join(input_directory, f"{storage_filename}.csv")):
+        logging.error(f"The storage file {storage_filename}.csv is missing!")
     else:
+        # hardcoded parameterlist of common parameters in all columns
```
>  # hardcoded parameterlist of common parameters in all columns

As this list is shortened, would there still be an error message if "crate" is not in the storage_xx.csv file, ie. for the input/output power?

Bachibouzouk commented:
No now I rebased #141 on your latest change (there was no conflicts \o/). Now you can write your tests  (write them at the bottom of the file to not create conflicts :).

You might want to have a look at the file in the branch `feature/test_A1` beforehand to see what changed
